phpcs ActionScheduler_Abstract_ListTable fixes

ovidiul opened the pull request:
This PR will address the PHPCS issues mentioned here https://github.com/woocommerce/action-scheduler/issues/660

Current PHPCS scan shows:

```

FILE: .../action-scheduler/classes/abstracts/ActionScheduler_Abstract_ListTable.php
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 7 ERRORS AND 2 WARNINGS AFFECTING 7 LINES
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
   1 | ERROR   | Missing file doc comment
 390 | WARNING | Comment refers to a TODO task
 406 | ERROR   | Use placeholders and $wpdb->prepare(); found $column
 406 | ERROR   | SQL wildcards for a LIKE query should be passed in through a replacement parameter and the variable part of the replacement should be escaped using "esc_like()". Found:  like "%%s%".
 406 | WARNING | Replacement variables found, but no valid placeholders found in the query.
 429 | ERROR   | Use placeholders and $wpdb->prepare(); found interpolated variable $column at "`$column` = %s"
 454 | ERROR   | Detected usage of a possibly undefined superglobal array index: $_SERVER['REQUEST_URI']. Use isset() or empty() to check the index exists before using it
 479 | ERROR   | Use placeholders and $wpdb->prepare(); found $sql
 482 | ERROR   | Use placeholders and $wpdb->prepare(); found $query_count
```

would it make sense to `phpcs:ignore` those?

barryhughes commented:
> would it make sense to phpcs:ignore those?

I think in many cases that would be fine; for some like `Missing file doc comment` I wonder if instead we should modify `phpcs.xml` to ignore that rule everywhere.

> Comment refers to a TODO task

For this one, thinking it is probably better that we don't ignore it (so that we get a reminder to take action) _or_ remove it from the source and create an issue for it instead.

ovidiul commented:
I've added the `Squiz.Commenting.FileComment.Missing` rules to the phpcs.xml file, although maybe I should open a separate PR for that?

Also added the comments for the rest of the items, and also refactored this [query](https://github.com/woocommerce/action-scheduler/pull/763/files#diff-18ffca2c4125958b98cb71171cb6ddfc0ec7829f127c0094e24c6c2796cce326R403-R405) for better handling.

As for the TODO, I've opened a new issue here https://github.com/woocommerce/action-scheduler/issues/766 .

Latest PHPCS shows no Warnings/Errors.

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -99,14 +126,18 @@ abstract class ActionScheduler_Abstract_ListTable extends WP_List_Table {
 	 * Makes translation easier, it basically just wraps
 	 * `_x` with some default (the package name).
 	 *
-	 * @deprecated 3.0.0
```
Was this dropped accidentally? Seems like we'd want to preserve it and I don't think it breaks our ruleset.

ovidiul replied:
Yes, my bad. It's back now, I've added the text `Use `__()` instead.` along with it, does it make sense to be there?

barryhughes replied:
The method description suggests it is a wrapper around `_x()` (even though it isn't) and, like `_x()` it accepts a `$context`, so _probably_ this would be just a little better:

```
* @deprecated 3.0.0 Use `_x()` instead.
```

ovidiul replied:
Makes sense, changed text added.

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -343,21 +387,22 @@ protected function get_table_columns() {
 		foreach ( $this->search_by as $column ) {
-			$filter[] = $wpdb->prepare('`' . $column . '` like "%%s%"', $wpdb->esc_like( $_GET['s'] ));
+			$wild     = '%';
+			$sql_like = $wild . $wpdb->esc_like( $search_string ) . $wild;
```
Do we need to pass `$search_string` through `esc_like()` twice (looking up at like 399)?

ovidiul replied:
Of course not, somehow I've overseen this. I've left the esc_like in the SQL statement as it makes more sense there and removed the duplicate above.

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -574,7 +655,7 @@ protected function display_header() {
 	protected function display_admin_notices() {
 		foreach ( $this->admin_notices as $notice ) {
-			echo '<div id="message" class="' . $notice['class'] . '">';
+			echo '<div id="message" class="' . esc_html( $notice['class'] ) . '">';
```
`esc_attr()` could be a better choice here, _or_ `sanitize_html_class()` (but the latter would only be safe if we are certain multiple classes are not being passed in here):

```suggestion
			echo '<div id="message" class="' . esc_attr( $notice['class'] ) . '">';
```

ovidiul replied:
Suggestion added :)

barryhughes reviewed (commented):
> I've added the Squiz.Commenting.FileComment.Missing rules to the phpcs.xml file, although maybe I should open a separate PR for that?

Perhaps, but since you've added it here let's just keep it in one place (thanks for adding it!).

barryhughes reviewed (changes requested):
Left a couple of minor observations: beyond that I just want to do a bit of testing before we merge (changes look great, but it's also a lot of change in a key class—so I just want to make sure we're covered).

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -612,7 +695,7 @@ protected function display_filter_by_status() {
 		if ( $status_list_items ) {
 			echo '<ul class="subsubsub">';
-			echo implode( " | \n", $status_list_items );
+			echo implode( " | \n", esc_html( $status_list_items ) );
```
This triggers an error, breaking the Tools → Scheduled Actions screen entirely (because `implode` expects the second param to be an array—and passing the array through `esc_html()` gives us a string).

Each element is already [composed in a safe manner](https://github.com/woocommerce/action-scheduler/pull/763/files#diff-18ffca2c4125958b98cb71171cb6ddfc0ec7829f127c0094e24c6c2796cce326R693) and contains HTML, so `esc_html()` wouldn't be a good fit (since it would expose the HTML). 

Alternative:

```suggestion
			echo implode( " | \n", $status_list_items ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -146,22 +183,31 @@ protected function process_bulk_action() {
 		if ( array_key_exists( $action, $this->bulk_actions ) && is_callable( array( $this, $method ) ) && ! empty( $_GET['ID'] ) && is_array( $_GET['ID'] ) ) {
 			$ids_sql = '(' . implode( ',', array_fill( 0, count( $_GET['ID'] ), '%s' ) ) . ')';
-			$this->$method( $_GET['ID'], $wpdb->prepare( $ids_sql, $_GET['ID'] ) );
+			$id      = sanitize_text_field( wp_unslash( $_GET['ID'] ) );
```
`$_GET['ID']` will be an array (we're dealing with bulk actions, so more than one can be passed), so we need to take a different approach as passing an array through `sanitize_text_field()` gives us an empty string and that means the bulk action will not be processed successfully.

Recommend:

```suggestion
			$id      = array_map( 'absint', $_GET['ID'] );
```

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -527,32 +593,49 @@ protected function maybe_render_actions( $row, $column_name ) {
 		$parameters = array( 'row_action', 'row_id', 'nonce' );
 		foreach ( $parameters as $parameter ) {
-			if ( empty( $_REQUEST[ $parameter ] ) ) {
+			if ( ! isset( $_REQUEST[ $parameter ] ) || empty( $_REQUEST[ $parameter ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Just curious, why the extra check here? I don't think it's needed to satisfy our linting rules (and I think `isset` and `empty` both have similar performance characteristics in this scenario—where the element is missing entirely, I mean)?

ovidiul replied:
definitely not needed, `isset` check removed now

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -527,32 +593,49 @@ protected function maybe_render_actions( $row, $column_name ) {
 
-		if ( $_REQUEST['nonce'] === wp_create_nonce( $_REQUEST[ 'row_action' ] . '::' . $_REQUEST[ 'row_id' ] ) && method_exists( $this, $method ) ) {
-			$this->$method( $_REQUEST['row_id'] );
+		if ( wp_create_nonce( $action . '::' . $row_id ) === $nonce && method_exists( $this, $method ) ) {
```
I realize you didn't introduce this and are just making a change to satisfy the yoda rule, but thinking about it this isn't the right way to check a nonce.

*Optional* thing we can slip in:

```suggestion
		if ( wp_verify_nonce( $nonce, $action . '::' . $row_id ) && method_exists( $this, $method ) ) {
```

barryhughes commented on classes/abstracts/ActionScheduler_Abstract_ListTable.php:
```diff
@@ -644,9 +727,9 @@ public function process_actions() {
 		$this->process_row_actions();
 
-		if ( ! empty( $_REQUEST['_wp_http_referer'] ) ) {
+		if ( ! empty( $_REQUEST['_wp_http_referer'] && ! empty( $_SERVER['REQUEST_URI'] ) ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Order of parentheses needs to be tweaked:

```suggestion
		if ( ! empty( $_REQUEST['_wp_http_referer'] ) && ! empty( $_SERVER['REQUEST_URI'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```

barryhughes reviewed (changes requested):
Sorry for the drip feed of notes here, but I think we'll be all set when the remaining points are addressed:

- [Change needed in the `get_bulk_actions()` method](https://github.com/woocommerce/action-scheduler/pull/763/files#r718788123)
- [Change needed in the `process_actions()` method](https://github.com/woocommerce/action-scheduler/pull/763/files#r718817923)
- [*Optional* change re nonce verification](https://github.com/woocommerce/action-scheduler/pull/763/files#r718811603)

ovidiul commented:
@barryhughes latest changes made, thank you for your guidance on this!

barryhughes reviewed (approved):
Great, thank you!

barryhughes commented:
Tests are failing, but very probably because of [this change](https://make.wordpress.org/core/2021/09/27/changes-to-the-wordpress-core-php-test-suite/) rather than anything in this PR.

barryhughes commented:
Thanks again!

Merged via https://github.com/woocommerce/action-scheduler/pull/779 (which was just to integrate changes from `master`, in order to have our CI checks run against this changeset).
